apps/aecore/lib/aecore/peers/sync.ex

```diff
@@ -15,30 +17,89 @@ defmodule Aecore.Peers.Sync do
   end
 
   def init(state) do
-    Process.send_after(self(), :work, 5_000)
     {:ok, state}
   end
 
-  def handle_info(:work, state) do
-    check_peers()
-    introduce_variety()
-    refill()
-    schedule_work()
-    {:noreply, state}
+  @spec add_block_to_state(binary(), term()) :: :ok
+  def add_block_to_state(block_hash, peer) do
+    GenServer.call(__MODULE__, {:add_block_to_state, block_hash, peer})
   end
 
-  defp schedule_work() do
-    Process.send_after(self(), :work, @check_time)
+  @spec ask_peers_for_unknown_blocks(map()) :: :ok
+  def ask_peers_for_unknown_blocks(peers) do
+    GenServer.call(__MODULE__, {:ask_peers_for_unknown_blocks, peers})
   end
 
-  defp check_peers do
-    Peers.check_peers()
+  @spec update_statuses() :: :ok
+  def update_statuses() do
+    GenServer.call(__MODULE__, :update_statuses)
+  end
+
+  @spec single_validate_all_blocks() :: :ok
+  def single_validate_all_blocks() do
+    GenServer.call(__MODULE__, :single_validate_all_blocks)
+  end
+
+  @spec add_valid_peer_blocks_to_chain() :: :ok
+  def add_valid_peer_blocks_to_chain() do
+    GenServer.call(__MODULE__, :add_valid_peer_blocks_to_chain)
+  end
+
+  def handle_call({:add_block_to_state, block_hash, peer}, _from, state) do
+    updated_state = Map.put(state, block_hash, %{peer: peer, status: :bad})
+    {:reply, :ok, updated_state}
+  end
+
+  def handle_call({:ask_peers_for_unknown_blocks, peers}, _from, state) do
+    state = Enum.reduce(peers, state, fn ({uri, latest_block_hash}, acc) ->
+        Map.merge(acc, check_peer_block(uri, latest_block_hash, %{}))
+      end)
+
+    {:reply, :ok, state}
+  end
+
+  def handle_call(:update_statuses, _from, state) do
+    updated_state = for {block_hash, %{peer: peer, status: status}} <- state, into: %{} do
+      case Chain.get_block_by_hex_hash(block_hash) do
+        {:error, _} ->
+          {block_hash, %{peer: peer, status: status}}
+        block
+         ->
+          must_be_updated = status == :bad && Map.has_key?(state, block.header.prev_hash)
+          case must_be_updated do
+            true ->
+              {block_hash, %{peer: peer, status: :good}}
+            false ->
+              {block_hash, %{peer: peer, status: status}}
+          end
+      end
+    end
+
+    {:reply, :ok, updated_state}
+  end
+
+  def handle_call(:single_validate_all_blocks, _from ,state) do
+    updated_state = single_validate_all_blocks(state)
+
+    {:reply, :ok, updated_state}
+  end
+
+  def handle_call(:add_valid_peer_blocks_to_chain, _from, state) do
+    updated_state = single_validate_all_blocks(state)
+
+    filtered_state = Enum.reduce(updated_state, updated_state, fn({block_hash, %{peer: peer}}, acc) ->
+          built_chain = build_chain(acc, {block_hash, peer}, [])
+          add_built_chain(built_chain)
+          remove_added_blocks_from_state(acc)
```

Comment:
We should combine this two functions into one (`add_built_chain` and `remove_added_blocks_from_state`). Going though whole `acc` every time can be time consuming.